File: cloudflare/game-contract.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import test from "node:test";
import { fileURLToPath } from "node:url";

import {
	COMPLETION_BONUS,
	GameError,
	MAX_PLAYER_NAME_CODE_POINTS,
	MAX_SCORE_CORRECTION_DELTA_MAGNITUDE,
	MAX_TOPIC_CODE_POINTS,
	MAX_TURN_ID_NUMBER,
	MINIMUM_SCORE,
	REMOTE_CLAIM_OBSERVATION_TOLERANCE_SECONDS,
	REMOTE_COMPLETION_GRACE_SECONDS,
	TURN_ID_EXHAUSTED_SENTINEL,
	TURN_ID_PREFIX,
	applyAction,
	createRoomState,
	joinRoom,
	type RoomState,
	type Turn,
} from "./game.ts";

interface RepeatedText {
	repeat: string;
	count: number;
	suffix?: string;
	suffixUtf16?: number[];
}

interface GameContract {
	schemaVersion: number;
	constants: {
		textLengthUnit: string;
		maxPlayerNameCodePoints: number;
		maxTopicCodePoints: number;
		remoteClaimObservationToleranceSeconds: number;
		remoteCompletionGraceSeconds: number;
		completionBonus: number;
		maxScoreCorrectionDeltaMagnitude: number;
		minimumScore: number;
		turnIdPrefix: string;
		maxTurnIdNumber: number;
		turnIdExhaustedSentinel: number;
	};
	cases: {
		unicodeTruncation: Array<{
			id: string;
			field: "player_name" | "topic";
			input: RepeatedText;
			expected: RepeatedText;
			expectedCodePoints: number;
		}>;
		remoteTurnClaims: Array<{
			id: string;
			durationSeconds: number;
			claimedSpokenSeconds: number;
			requestedCompleted: boolean;
			eliminated: boolean;
			serverElapsedSeconds: number;
			expectedSpokenSeconds: number;
			expectedCompleted: boolean;
		}>;
		turnSubmissions: Array<{
			id: string;
			durationSeconds: number;
			spokenSeconds: number;
			requestedCompleted: boolean;
			eliminated: boolean;
			expectedSpokenSeconds: number;
			expectedCompleted: boolean;
			expectedScore: number;
		}>;
		scoreCorrections: Array<{
			id: string;
			playerId: string;
			initialScore: number;
			requestedDelta: number;
			expectedAppliedDelta: number;
			expectedScore: number;
			expectedAccepted: boolean;
			errorCode?: string;
			expectedVersionChanged?: boolean;
		}>;
		customTopics: Array<{
			id: string;
			inputTopics: string[];
			initial: {
				topics: string[];
				topicPack: string;
				deck: number[];
				deckCursor: number;
				topicGeneration: number;
				version: number;
			};
			expected: {
				topics: string[];
				topicPack: string;
				deck: number[];
				deckCursor: number;
				topicGeneration: number;
				version: number;
				accepted: boolean;
				errorCode: string;
			};
		}>;
		turnCounters: Array<{
			id: string;
			allocation: "start_turn" | "redraw_turn";
			initialNextTurn: number;
			activeTurnId: string | null;
			completedTurnIds: string[];
			expectedAllocatedTurnId: string | null;
			expectedNextTurn: number;
			expectedAccepted: boolean;
			errorCode?: string;
			expectedHistoryPreserved?: boolean;
			expectedVersionChanged?: boolean;
		}>;
	};
}

const contract = JSON.parse(
	readFileSync(join(dirname(fileURLToPath(import.meta.url)), "../testdata/game-contract.v1.json"), "utf8"),
) as GameContract;

const host = "a".repeat(64);
const guest = "b".repeat(64);

function expand(value: RepeatedText): string {
	const suffix = value.suffixUtf16
		? String.fromCharCode(...value.suffixUtf16)
		: value.suffix ?? "";
	return value.repeat.repeat(value.count) + suffix;
}

function roomWithTwoPlayers(duration = 60): RoomState {
	const room = createRoomState("ABC234", host, "Alice", 0);
	joinRoom(room, guest, "Bob", 1);
	applyAction(room, host, { type: "settings", duration }, 2);
	return room;
}

function turnWithId(room: RoomState, id: string): Turn {
	const player = room.players[0];
	return {
		id,
		playerId: player.id,
		playerName: player.name,
		round: 1,
		topic: room.topics[0],
		topicIndex: 0,
		duration: room.settings.duration,
		silence: room.settings.silence,
		begunAt: null,
	};
}

function errorCode(error: unknown): string | null {
	if (!(error instanceof GameError)) return null;
	if (error.status === 404 && error.message === "Player not found.") return "player_not_found";
	if (error.message === "Choose at least one topic.") return "topics_required";
	if (error.status === 409 && error.message === "This room has exhausted its turn IDs.") return "turn_ids_exhausted";
	return null;
}

test("shared game contract constants match the Cloudflare engine", () => {
	assert.equal(contract.schemaVersion, 1);
	assert.equal(contract.constants.textLengthUnit, "unicode-code-points");
	assert.equal(contract.constants.maxPlayerNameCodePoints, MAX_PLAYER_NAME_CODE_POINTS);
	assert.equal(contract.constants.maxTopicCodePoints, MAX_TOPIC_CODE_POINTS);
	assert.equal(contract.constants.remoteClaimObservationToleranceSeconds, REMOTE_CLAIM_OBSERVATION_TOLERANCE_SECONDS);
	assert.equal(contract.constants.remoteCompletionGraceSeconds, REMOTE_COMPLETION_GRACE_SECONDS);
	assert.equal(contract.constants.completionBonus, COMPLETION_BONUS);
	assert.equal(contract.constants.maxScoreCorrectionDeltaMagnitude, MAX_SCORE_CORRECTION_DELTA_MAGNITUDE);
	assert.equal(contract.constants.minimumScore, MINIMUM_SCORE);
	assert.equal(contract.constants.turnIdPrefix, TURN_ID_PREFIX);
	assert.equal(contract.constants.maxTurnIdNumber, MAX_TURN_ID_NUMBER);
	assert.equal(contract.constants.turnIdExhaustedSentinel, TURN_ID_EXHAUSTED_SENTINEL);
});

test("shared Unicode truncation cases use whole code points", () => {
	for (const fixture of contract.cases.unicodeTruncation) {
		const input = expand(fixture.input);
		const expected = expand(fixture.expected);
		const room = createRoomState("ABC234", host, fixture.field === "player_name" ? input : "Alice", 0);
		let actual: string;
		if (fixture.field === "topic") {
			applyAction(room, host, { type: "custom-topics", topics: [input] }, 1);
			actual = room.topics[0];
		} else {
			actual = room.players[0].name;
		}
		assert.equal(actual, expected, fixture.id);
		assert.equal([...actual].length, fixture.expectedCodePoints, fixture.id);
	}
});

test("shared remote turn claims use observation tolerance and completion grace", () => {
	for (const fixture of contract.cases.remoteTurnClaims) {
		const room = roomWithTwoPlayers(fixture.durationSeconds);
		applyAction(room, host, { type: "move-player", playerId: "p2", offset: -1 }, 3);
		applyAction(room, host, { type: "start-game" }, 4);
		applyAction(room, guest, { type: "start-turn", afterTurnId: "" }, 5);
		const turnId = room.activeTurn?.id;
		const begunAt = 100_000;
		if (fixture.serverElapsedSeconds >= 0) {
			applyAction(room, guest, { type: "begin-turn", turnId }, begunAt);
		}
		applyAction(room, guest, {
			type: "submit-turn",
			turnId,
			spokenSeconds: fixture.claimedSpokenSeconds,
			completed: fixture.requestedCompleted,
			eliminated: fixture.eliminated,
		}, begunAt + Math.max(0, fixture.serverElapsedSeconds) * 1_000);
		const completed = room.completedTurns[0];
		assert.equal(completed.spokenSeconds, fixture.expectedSpokenSeconds, fixture.id);
		assert.equal(completed.completed, fixture.expectedCompleted, fixture.id);
	}
});

test("shared turn submission cases normalize completion and scoring", () => {
	for (const fixture of contract.cases.turnSubmissions) {
		const room = roomWithTwoPlayers(fixture.durationSeconds);
		applyAction(room, host, { type: "start-game" }, 3);
		applyAction(room, host, { type: "start-turn", afterTurnId: "" }, 4);
		applyAction(room, host, {
			type: "submit-turn",
			turnId: room.activeTurn?.id,
			spokenSeconds: fixture.spokenSeconds,
			completed: fixture.requestedCompleted,
			eliminated: fixture.eliminated,
		}, 5);
		const completed = room.completedTurns[0];
		assert.equal(completed.spokenSeconds, fixture.expectedSpokenSeconds, fixture.id);
		assert.equal(completed.completed, fixture.expectedCompleted, fixture.id);
		assert.equal(completed.score, fixture.expectedScore, fixture.id);
	}
});

test("shared score correction cases enforce bounds, floor, and rejection", () => {
	for (const fixture of contract.cases.scoreCorrections) {
		const room = createRoomState("ABC234", host, "Alice", 0);
		room.players[0].score = fixture.initialScore;
		const initialVersion = room.version;
		let caught: unknown;
		try {
			applyAction(room, host, {
				type: "score",
				playerId: fixture.playerId,
				delta: fixture.requestedDelta,
			}, 1);
		} catch (error) {
			caught = error;
		}

		assert.equal(caught === undefined, fixture.expectedAccepted, fixture.id);
		assert.equal(room.players[0].score, fixture.expectedScore, fixture.id);
		const boundedDelta = fixture.expectedAccepted
			? Math.max(
				-contract.constants.maxScoreCorrectionDeltaMagnitude,
				Math.min(contract.constants.maxScoreCorrectionDeltaMagnitude, fixture.requestedDelta),
			)
			: 0;
		assert.equal(boundedDelta, fixture.expectedAppliedDelta, fixture.id);
		if (fixture.errorCode) assert.equal(errorCode(caught), fixture.errorCode, fixture.id);
		if (fixture.expectedVersionChanged !== undefined) {
			assert.equal(room.version !== initialVersion, fixture.expectedVersionChanged, fixture.id);
		}
		assert.ok(room.players[0].score >= contract.constants.minimumScore, fixture.id);
	}
});

test("shared empty custom-topic case rejects atomically", () => {
	for (const fixture of contract.cases.customTopics) {
		const room = createRoomState("ABC234", host, "Alice", 0);
		room.topics = [...fixture.initial.topics];
		room.settings.topicPack = fixture.initial.topicPack;
		room.deck = [...fixture.initial.deck];
		room.deckCursor = fixture.initial.deckCursor;
		room.topicGeneration = fixture.initial.topicGeneration;
		room.version = fixture.initial.version;
		let caught: unknown;
		try {
			applyAction(room, host, { type: "custom-topics", topics: fixture.inputTopics }, 1);
		} catch (error) {
			caught = error;
		}

		assert.equal(caught === undefined, fixture.expected.accepted, fixture.id);
		assert.equal(errorCode(caught), fixture.expected.errorCode, fixture.id);
		assert.deepEqual(room.topics, fixture.expected.topics, fixture.id);
		assert.equal(room.settings.topicPack, fixture.expected.topicPack, fixture.id);
		assert.deepEqual(room.deck, fixture.expected.deck, fixture.id);
		assert.equal(room.deckCursor, fixture.expected.deckCursor, fixture.id);
		assert.equal(room.topicGeneration, fixture.expected.topicGeneration, fixture.id);
		assert.equal(room.version, fixture.expected.version, fixture.id);
	}
});

test("shared stale turn-counter cases repair lazily without rewriting history", () => {
	for (const fixture of contract.cases.turnCounters) {
		const room = roomWithTwoPlayers();
		room.nextTurn = fixture.initialNextTurn;
		room.completedTurns = fixture.completedTurnIds.map((id) => turnWithId(room, id));
		const initialVersion = room.version;
		let caught: unknown;
		try {
			if (fixture.allocation === "start_turn") {
				applyAction(room, host, { type: "start-game" }, 3);
				applyAction(room, host, { type: "start-turn", afterTurnId: "" }, 4);
			} else {
				applyAction(room, host, { type: "start-game" }, 3);
				room.nextTurn = fixture.initialNextTurn;
				room.completedTurns = fixture.completedTurnIds.map((id) => turnWithId(room, id));
				room.activeTurn = turnWithId(room, fixture.activeTurnId ?? "");
				applyAction(room, host, { type: "redraw-turn", turnId: fixture.activeTurnId }, 4);
			}
		} catch (error) {
			caught = error;
		}

		assert.equal(caught === undefined, fixture.expectedAccepted, fixture.id);
		if (fixture.errorCode) assert.equal(errorCode(caught), fixture.errorCode, fixture.id);
		assert.equal(room.activeTurn?.id ?? null, fixture.expectedAllocatedTurnId, fixture.id);
		assert.equal(room.nextTurn, fixture.expectedNextTurn, fixture.id);
		if (fixture.expectedVersionChanged !== undefined) {
			assert.equal(room.version !== initialVersion, fixture.expectedVersionChanged, fixture.id);
		}
		if (fixture.expectedHistoryPreserved) {
			assert.deepEqual(room.completedTurns.map((turn) => turn.id), fixture.completedTurnIds, fixture.id);
		}
		if (fixture.allocation === "redraw_turn") {
			assert.deepEqual(room.completedTurns.map((turn) => turn.id), fixture.completedTurnIds, fixture.id);
		}
		if (fixture.expectedAccepted) {
			assert.ok(room.activeTurn?.id.startsWith(contract.constants.turnIdPrefix), fixture.id);
		}
	}
});

File: internal/game/game_contract_test.go
package game

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"slices"
	"strings"
	"testing"
	"unicode/utf16"
	"unicode/utf8"
)

type gameContractFixture struct {
	SchemaVersion int `json:"schemaVersion"`
	Constants     struct {
		MaxPlayerNameCodePoints          int    `json:"maxPlayerNameCodePoints"`
		MaxTopicCodePoints               int    `json:"maxTopicCodePoints"`
		CompletionBonus                  int    `json:"completionBonus"`
		MaxScoreCorrectionDeltaMagnitude int    `json:"maxScoreCorrectionDeltaMagnitude"`
		MinimumScore                     int    `json:"minimumScore"`
		TurnIDPrefix                     string `json:"turnIdPrefix"`
		MaxTurnIDNumber                  int64  `json:"maxTurnIdNumber"`
		TurnIDExhaustedSentinel          int64  `json:"turnIdExhaustedSentinel"`
	} `json:"constants"`
	Cases struct {
		UnicodeTruncation []struct {
			ID    string `json:"id"`
			Field string `json:"field"`
			Input struct {
				Repeat      string   `json:"repeat"`
				Count       int      `json:"count"`
				Suffix      string   `json:"suffix"`
				SuffixUTF16 []uint16 `json:"suffixUtf16"`
			} `json:"input"`
			Expected struct {
				Repeat      string   `json:"repeat"`
				Count       int      `json:"count"`
				Suffix      string   `json:"suffix"`
				SuffixUTF16 []uint16 `json:"suffixUtf16"`
			} `json:"expected"`
			ExpectedCodePoints int `json:"expectedCodePoints"`
		} `json:"unicodeTruncation"`
		TurnSubmissions []struct {
			ID                    string `json:"id"`
			DurationSeconds       int    `json:"durationSeconds"`
			SpokenSeconds         int    `json:"spokenSeconds"`
			RequestedCompleted    bool   `json:"requestedCompleted"`
			Eliminated            bool   `json:"eliminated"`
			ExpectedSpokenSeconds int    `json:"expectedSpokenSeconds"`
			ExpectedCompleted     bool   `json:"expectedCompleted"`
			ExpectedScore         int    `json:"expectedScore"`
		} `json:"turnSubmissions"`
		ScoreCorrections []struct {
			ID                   string `json:"id"`
			PlayerID             string `json:"playerId"`
			InitialScore         int    `json:"initialScore"`
			RequestedDelta       int    `json:"requestedDelta"`
			ExpectedAppliedDelta int    `json:"expectedAppliedDelta"`
			ExpectedScore        int    `json:"expectedScore"`
			ExpectedAccepted     bool   `json:"expectedAccepted"`
			ErrorCode            string `json:"errorCode"`
		} `json:"scoreCorrections"`
		CustomTopics []struct {
			ID          string   `json:"id"`
			InputTopics []string `json:"inputTopics"`
			Initial     struct {
				Topics     []string `json:"topics"`
				TopicPack  string   `json:"topicPack"`
				Deck       []int    `json:"deck"`
				DeckCursor int      `json:"deckCursor"`
			} `json:"initial"`
			Expected struct {
				Topics     []string `json:"topics"`
				TopicPack  string   `json:"topicPack"`
				Deck       []int    `json:"deck"`
				DeckCursor int      `json:"deckCursor"`
				Accepted   bool     `json:"accepted"`
				ErrorCode  string   `json:"errorCode"`
			} `json:"expected"`
		} `json:"customTopics"`
		TurnCounters []struct {
			ID                       string   `json:"id"`
			Allocation               string   `json:"allocation"`
			InitialNextTurn          int64    `json:"initialNextTurn"`
			ActiveTurnID             *string  `json:"activeTurnId"`
			CompletedTurnIDs         []string `json:"completedTurnIds"`
			ExpectedAllocatedTurnID  *string  `json:"expectedAllocatedTurnId"`
			ExpectedNextTurn         int64    `json:"expectedNextTurn"`
			ExpectedAccepted         bool     `json:"expectedAccepted"`
			ErrorCode                string   `json:"errorCode"`
			ExpectedHistoryPreserved bool     `json:"expectedHistoryPreserved"`
		} `json:"turnCounters"`
	} `json:"cases"`
}

func loadGameContract(t *testing.T) gameContractFixture {
	t.Helper()
	path := filepath.Join("..", "..", "testdata", "game-contract.v1.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read game contract: %v", err)
	}
	var fixture gameContractFixture
	if err := json.Unmarshal(data, &fixture); err != nil {
		t.Fatalf("decode game contract: %v", err)
	}
	if fixture.SchemaVersion != 1 {
		t.Fatalf("unsupported game contract schema %d", fixture.SchemaVersion)
	}
	return fixture
}

func expandContractText(repeat string, count int, suffix string, suffixUTF16 []uint16) string {
	if len(suffixUTF16) > 0 {
		suffix = string(utf16.Decode(suffixUTF16))
	}
	return strings.Repeat(repeat, count) + suffix
}

func TestGameContractUnicodeTruncation(t *testing.T) {
	contract := loadGameContract(t)
	if contract.Constants.MaxPlayerNameCodePoints != MaxPlayerNameLength ||
		contract.Constants.MaxTopicCodePoints != MaxTopicLength {
		t.Fatal("runtime text limits differ from the shared contract")
	}

	for _, test := range contract.Cases.UnicodeTruncation {
		t.Run(test.ID, func(t *testing.T) {
			input := expandContractText(test.Input.Repeat, test.Input.Count, test.Input.Suffix, test.Input.SuffixUTF16)
			expected := expandContractText(test.Expected.Repeat, test.Expected.Count, test.Expected.Suffix, test.Expected.SuffixUTF16)
			var actual string
			session := NewSession("contract")
			switch test.Field {
			case "player_name":
				actual = session.AddPlayer(input).Name
			case "topic":
				if err := session.SetTopics([]string{input}); err != nil {
					t.Fatal(err)
				}
				actual = session.Topics[0]
			default:
				t.Fatalf("unsupported field %q", test.Field)
			}
			if actual != expected || utf8.RuneCountInString(actual) != test.ExpectedCodePoints {
				t.Fatalf("got %q (%d code points), want %q (%d)", actual,
					utf8.RuneCountInString(actual), expected, test.ExpectedCodePoints)
			}
		})
	}
}

func TestGameContractGoInputNormalizesMalformedUTF8(t *testing.T) {
	// JSON cannot carry malformed UTF-8, so this Go-only adapter check proves
	// the same scalar-value guarantee that toWellFormed enforces in JavaScript.
	session := NewSession("contract")
	player := session.AddPlayer(string([]byte{0xff}))
	if player.Name != "\uFFFD" || !utf8.ValidString(player.Name) {
		t.Fatalf("malformed name normalized to %q", player.Name)
	}
	if err := session.SetTopics([]string{string([]byte{0xff})}); err != nil {
		t.Fatal(err)
	}
	if len(session.Topics) != 1 || session.Topics[0] != "\uFFFD" || !utf8.ValidString(session.Topics[0]) {
		t.Fatalf("malformed topic normalized to %q", session.Topics)
	}
}

func TestGameContractTurnSubmissions(t *testing.T) {
	contract := loadGameContract(t)
	if contract.Constants.CompletionBonus != CompletionBonus {
		t.Fatal("runtime completion bonus differs from the shared contract")
	}
	for _, test := range contract.Cases.TurnSubmissions {
		t.Run(test.ID, func(t *testing.T) {
			session := NewSession("contract")
			session.AddPlayer("Avery")
			session.AddPlayer("Blair")
			session.Settings.SpeakingDurationSeconds = test.DurationSeconds
			if err := session.SetTopics([]string{"Contract topic"}); err != nil {
				t.Fatal(err)
			}
			if _, err := session.StartTurn(); err != nil {
				t.Fatal(err)
			}
			turn, err := session.SubmitTurn(test.SpokenSeconds, test.RequestedCompleted, test.Eliminated)
			if err != nil {
				t.Fatal(err)
			}
			if turn.SpokenSeconds != test.ExpectedSpokenSeconds ||
				turn.Completed != test.ExpectedCompleted || turn.Score != test.ExpectedScore {
				t.Fatalf("got spoken=%d completed=%v score=%d, want %d %v %d", turn.SpokenSeconds,
					turn.Completed, turn.Score, test.ExpectedSpokenSeconds, test.ExpectedCompleted, test.ExpectedScore)
			}
		})
	}
}

func TestGameContractScoreCorrections(t *testing.T) {
	contract := loadGameContract(t)
	if contract.Constants.MaxScoreCorrectionDeltaMagnitude != MaxScoreCorrectionDelta ||
		contract.Constants.MinimumScore != 0 {
		t.Fatal("runtime score limits differ from the shared contract")
	}
	for _, test := range contract.Cases.ScoreCorrections {
		t.Run(test.ID, func(t *testing.T) {
			session := NewSession("contract")
			player := session.AddPlayer("Avery")
			session.Players[0].Score = test.InitialScore
			if player.ID != "p1" {
				t.Fatalf("fixture requires p1, got %q", player.ID)
			}
			applied, accepted := session.OverrideScore(test.PlayerID, test.RequestedDelta)
			if applied != test.ExpectedAppliedDelta || accepted != test.ExpectedAccepted ||
				session.Players[0].Score != test.ExpectedScore {
				t.Fatalf("got applied=%d accepted=%v score=%d, want %d %v %d", applied, accepted,
					session.Players[0].Score, test.ExpectedAppliedDelta, test.ExpectedAccepted, test.ExpectedScore)
			}
			if !accepted && test.ErrorCode != "player_not_found" {
				t.Fatalf("unsupported rejection code %q", test.ErrorCode)
			}
		})
	}
}

func TestGameContractEmptyTopicsAreAtomic(t *testing.T) {
	contract := loadGameContract(t)
	for _, test := range contract.Cases.CustomTopics {
		t.Run(test.ID, func(t *testing.T) {
			session := NewSession("contract")
			session.Topics = append([]string(nil), test.Initial.Topics...)
			session.Settings.TopicPackID = test.Initial.TopicPack
			session.TopicDeck = append([]int(nil), test.Initial.Deck...)
			session.TopicDeckCursor = test.Initial.DeckCursor

			err := session.SetTopics(test.InputTopics)
			accepted := err == nil
			if accepted != test.Expected.Accepted {
				t.Fatalf("accepted=%v, want %v", accepted, test.Expected.Accepted)
			}
			if test.Expected.ErrorCode != "topics_required" || !errors.Is(err, ErrTopicsRequired) {
				t.Fatalf("got error %v for code %q", err, test.Expected.ErrorCode)
			}
			if !reflect.DeepEqual(session.Topics, test.Expected.Topics) ||
				session.Settings.TopicPackID != test.Expected.TopicPack ||
				!reflect.DeepEqual(session.TopicDeck, test.Expected.Deck) ||
				session.TopicDeckCursor != test.Expected.DeckCursor {
				t.Fatalf("rejected replacement mutated session: %+v", session)
			}
		})
	}
}

func TestGameContractTurnCounterRepair(t *testing.T) {
	contract := loadGameContract(t)
	if contract.Constants.TurnIDPrefix != "t" {
		t.Fatalf("unsupported turn prefix %q", contract.Constants.TurnIDPrefix)
	}
	if contract.Constants.MaxTurnIDNumber != MaxTurnIDNumber ||
		contract.Constants.TurnIDExhaustedSentinel != turnIDExhaustedSentinel {
		t.Fatal("runtime turn-ID range differs from the shared contract")
	}
	for _, test := range contract.Cases.TurnCounters {
		t.Run(test.ID, func(t *testing.T) {
			session := NewSession("contract")
			session.AddPlayer("Avery")
			session.AddPlayer("Blair")
			if err := session.SetTopics([]string{"First", "Second"}); err != nil {
				t.Fatal(err)
			}
			session.NextTurnNumber = test.InitialNextTurn
			for _, id := range test.CompletedTurnIDs {
				session.CompletedTurns = append(session.CompletedTurns, Turn{ID: id})
			}

			var allocated *Turn
			var err error
			switch test.Allocation {
			case "start_turn":
				allocated, err = session.StartTurn()
			case "redraw_turn":
				if test.ActiveTurnID == nil {
					t.Fatal("redraw fixture requires an active turn")
				}
				session.Started = true
				session.ActiveTurn = &Turn{ID: *test.ActiveTurnID, PlayerID: "p1", PlayerName: "Avery", Topic: "First", TopicIndex: 0}
				allocated, err = session.RedrawActiveTurn()
			default:
				t.Fatalf("unsupported allocation %q", test.Allocation)
			}
			accepted := err == nil
			if accepted != test.ExpectedAccepted {
				t.Fatalf("accepted=%v with error %v, want %v", accepted, err, test.ExpectedAccepted)
			}
			if test.ErrorCode == "turn_ids_exhausted" {
				if !errors.Is(err, ErrTurnIDsExhausted) {
					t.Fatalf("got error %v, want ErrTurnIDsExhausted", err)
				}
			} else if err != nil {
				t.Fatal(err)
			}
			var allocatedID *string
			if allocated != nil {
				id := allocated.ID
				allocatedID = &id
			}
			if !reflect.DeepEqual(allocatedID, test.ExpectedAllocatedTurnID) ||
				session.NextTurnNumber != test.ExpectedNextTurn {
				t.Fatalf("got ID=%v next=%d, want %v %d", allocatedID, session.NextTurnNumber,
					test.ExpectedAllocatedTurnID, test.ExpectedNextTurn)
			}
			if test.ExpectedHistoryPreserved {
				gotIDs := make([]string, 0, len(session.CompletedTurns))
				for _, turn := range session.CompletedTurns {
					gotIDs = append(gotIDs, turn.ID)
				}
				if !slices.Equal(gotIDs, test.CompletedTurnIDs) {
					t.Fatalf("completed IDs changed to %v, want %v", gotIDs, test.CompletedTurnIDs)
				}
			}
		})
	}
}
